### PR TITLE
SR and PP Algorithm Overhaul

### PR DESCRIPTION
**Pull request by Xexxar**
# osu! Standard Star Rating Overhaul Proposal

## Introduction

Hello all,

This is my PR for a series of changes that will drastically improve the performance point algorithm going forward over the next years. In creating this PR. I am not stating that I think that this build is perfectly balanced, or that this is the end goal solution for PP, but rather I am saying that this is a large step forwards in creating an accurate rating system of the game's many challenging aspects. There are some exciting gains, and exciting loses, and I'm looking forward to explaining them.

### Why now

A perfect PP system does not exist. New players and skillsets and metas will come and go as we change the algorithm. What is "performant" is a constantly evolving and expanding definition, and therefore it makes little sense to hold back from bringing in change for the sake of variety and developing that meta game further. For the last 7 years, we have had a very similar meta of jumps and streams that has held exclusive dominance. This rework hopes to change some of that, to bring a different variety of maps into people's top plays.

### Core Philosophy

Generally with every rework I am a part of, the purpose of manipulating the algorithm isn't necessarily to try quantify the "greatest objective difficulty" (as this is literally impossible), but rather it is to try to measure the quality of players "performances". If you go to a concert and every band plays the exact same song, it doesn't matter how well they play, objectively "performance" would be quite boring. The same analogy applies for osu! No one wants to see the same 10 scores on every profile. It is for this reason that my goal with every rework is to make the scores that stand out the most worth the most, even at the expense of "objective difficulty". Of course, many scores that are unique carry with them their own unique form of objective difficulty. Usually though, these unique scores carry some aspect to them that isn't properly quantified by the system.

## OsuSkill.cs - The Core SR Overhaul

The core component of this PR is the SR overhaul. Similar to multiple previous unmerged PRs (see https://github.com/ppy/osu/pull/11514), the goal of this SR change is to remove object based length from the PP system and to instead properly weigh the entire map at one time. The existing algorithm sums up star rating using 

```cs
for (in in Chunks)
{
    SR += chunks[i] * (.9)^i //psuedocoded
}
```
and
```cs
            // Longer maps are worth more
            double lengthBonus = 0.95 + 0.4 * Math.Min(1.0, totalHits / 2000.0) +
                                 (totalHits > 2000 ? Math.Log10(totalHits / 2000.0) * 0.5 : 0.0);

            ppValue *= lengthBonus;
```

It is no secret that this system is flawed. It fails to capture the majority of the map in the star rating, and instead does its best total summation and leaves object count to do the rest. It incentivizes mappers to use spikes to inflate the star rating as well as make the rest as easy as possible (see https://osu.ppy.sh/beatmapsets/906786#osu/1911308 for an example).

### Doubling Length Should Increase SR

The core idea of the new SR calculator is based on the idea that doubling the length of a map while keeping the difficulty constant should increase the overall difficulty by a fixed percentage. I.E if you have a 6 star map and you copy paste the same map, it should be 6 * k, where k is some constant (like 1.05). A graph has been built below to demonstrate how this works in practice.

https://www.desmos.com/calculator/gk5ujznstf

In this graph, we have two groups of notes, one where every object has difficulty d_1, and one where every object has d_2. We can see that as we add notes to any group, the total SR goes up, and we can adjust the amount the SR goes up based on the star_bonus_per_double constant. This means that, as we add objects of varying difficulty, we can properly weight additional notes relative to the skill level of other notes.

NOTE: We made it so display star rating did not correlate to difficulty in this way. We used the same idea of a .9^n geometric sum for the display SR in order to preserve similar SR numbers to what players are used to. It also just seemed weird that really long maps that used to be 7* were suddenly 8.5*.

### Full Combo Time Matters

In principle, every map takes time to FC, and thus the time needed to full combo the map should have a bearing into the star rating of the map. This section of maths is pretty complicated, so I will leave this for joz to explain this since he wrote (and probably has already documented it somewhere).

The main thing that I did differently was I used a varying desired time to FC. This was done because I use high StarsPerDouble values (which increases the difficulty with length at a much higher rate), and I wanted to lower the influence of length in consistent maps. Since star rating is already larger by nature of the length, we avoid doubly rewarding length by increasing the time desired to FC. This is done here:

```cs
        private double expectedTargetTime(double totalDifficulty)
        {
            double targetTime = 0;

            for (int i=1;i<strains.Count;i++)
            {
                targetTime += Math.Min(2000, times[i] - times[i-1]) * (strains[i] / totalDifficulty);
            }

            return targetTime;
        }

...
...

            target_fc_time += 45 * Math.Max(0, expectedTargetTime(totalDifficulty) - 60000);
            // for every minute of straining time past 1 minute, add 45 mins to estimated time to FC.
```

## Skills Rework

One of the major things that I wanted to do with this rework is remove the idea of "time based strain". Geometrically decaying strain with respect to time may seem like a good idea but has a number of problems when it comes to balancing. One such issue is that the number of objects it takes to reach "peak" strain (that is, where strain added = strain decayed) varies. This is problematic because it becomes much more difficult to balance between different beats per minute. It is for this reason that I changed both skills to decay at a fixed rate per object (unless the objects are more than 500 ms apart, then we continue to decay so as to not hold strain over breaks, etc).

### OsuDifficultyHitObject

In order to improve our skills, some new attributes were added. A vector2 distance vector was added to allow vector mathematics to be used with things like angle calculations, etc. Additional, another core piece that was added was Flow and Snap Probabilities. In osu, you generally have two styles of aim. One style is snapping, where you distinctly move to each object and stop on the note (think a jump). The other is flowing, where you continuously move through multiple notes without stopping your cursor (think a stream). These predictions are then used allow us to handle these two types of aims differently. For example, if we have a high flow probability, we might consider acute angles more difficult. Likewise, a high snap probability might imply that aim is more difficult.

### Aim

One of the core problems with the existing system is the separation of flow aim and snap aim. Historically, this was necessary due to the chunking geometric SR sum preventing us from summing across all parts of the map. However, since that is removed, we are able to put both skills together. Furthermore, we can reward for the overlap of these two styles of aim, such as patterns of spaced triplets, which interlace both flow aim and snap aim.

#### Snap

```cs
            double strain = 0;

            currVector = Vector2.Multiply(currVector, (float)snapScaling(osuCurrObj.JumpDistance / 100));
            prevVector = Vector2.Multiply(prevVector, (float)snapScaling(osuPrevObj.JumpDistance / 100));

            var observedDistance = Vector2.Add(currVector, Vector2.Multiply(prevVector, (float)(0.35)));

            strain = observedDistance.Length * osuCurrObj.SnapProbability;

            strain *= Math.Min(osuCurrObj.StrainTime / (osuCurrObj.StrainTime - 20) , osuPrevObj.StrainTime / (osuPrevObj.StrainTime - 20));
            // buff high BPM slightly.

            return strain;
```

I believe the code speaks for itself here. Before we use any vectors, we scale them using a log (thats what snapScaling is). Distance does make aim hard, but Fitt's law tells us that aim does not have linear relationship with difficulty. In this PR, I did not use fitt's law, but rather a modified version. As far as the calculation itself: you have to make a full stop with your cursor when you snap. This motion produces a backwards force against the direction towards the note. To recognize this, we add the prior vector, scaled slightly, to the current vector. Next we adjust for BPM. It's that simple.

#### Flow

```cs
        {
            double strain = 0;

            var observedDistance = Vector2.Subtract(currVector, Vector2.Multiply(prevVector, (float)0.1));

            double prevAngularMomentumChange = Math.Abs(osuCurrObj.Angle * currVector.Length - osuPrevObj.Angle * prevVector.Length);
            double nextAngularMomentumChange = Math.Abs(osuCurrObj.Angle * currVector.Length - osuNextObj.Angle * nextVector.Length);

            double angularMomentumChange = Math.Sqrt(Math.Min(currVector.Length, prevVector.Length) * Math.Abs(nextAngularMomentumChange - prevAngularMomentumChange) / (2 * Math.PI));
            // buff for changes in angular momentum, but only if the momentum change doesnt equal the previous.

            double momentumChange = Math.Sqrt(Math.Max(0, prevVector.Length - currVector.Length) * Math.Min(currVector.Length, prevVector.Length));
            // reward for accelerative changes in momentum

            strain = osuCurrObj.FlowProbability * (observedDistance.Length
                                                    + Math.Max(momentumChange * (0.5 + 0.5 * osuPrevObj.FlowProbability),
                                                      angularMomentumChange * osuPrevObj.FlowProbability));

            strain *= Math.Min(osuCurrObj.StrainTime / (osuCurrObj.StrainTime - 10) , osuPrevObj.StrainTime / (osuPrevObj.StrainTime - 10));
            // buff high BPM slightly.
```

The code here is a bit more complex. We first do the same type of operation with the previous vector and the current vector, but we subtract instead of add (thereby penalizing perfectly linear motion). This is done with a very small multiplier. The logic here is that maintaining a consistent rotational angle is harder than a linear angle, and this is a way to accomplish that. After that, we calculate the previous and current rotational momentums and then find their difference. This is done to buff stream cuts, or changes in flow. Going from a curved stream to a straight stream has a point of inflection where the angular momentum has to change, and so we reward for it. I also buff for deceleration changes in regular linear momentum. This is done to rewards stream jumps, or just simple single -> triplet patterns. 

All of the extra buffs are then geometrically averaged with the minimum velocity vector between the current and previous. This is done to prevent buffing momentum changes on a full stop, or reward an angle changes where the notes are completely overlapping. Some of these buffs are then scaled with the previous objects flow probability for balancing sake. Multiply by BPM and that's the skill.

After that, we simply sum these two styles of aim with their respective multiplies, along with slider difficulty (which isn't too robust, so I won't explain it).

### Tap

I should have probably just kept this called Speed for the simplicity of the PR. I'll leave it as is for now and the devs can make the call. I like the name "tap" more, since it has to do with tapping and "speed" is ambiguous. 

Tap has some slightly complicated elements to it, which I will gloss over. The main new part that has been added to tap is the inclusion of a "rhythm difficulty". Ironically the hardest thing to quantify in a rhythm game is rhythm. With simply having a historical view of StrainTimes, this is a very challenging problem to try and tackle. 

The idea of the rhythm calculator I built is to reward tapping if it has dense patterns of triplets, quads, doubles, etc. The way I do this is by summing the number of "islands" or groups of notes of these sizes, and the geometric average of the two notes associated with the transition into the island... For example, if you have 

`x___x_x_x___x_x_x____x_x___x_x`

we would calculate 2 island of size 3 and one island of size 2. We don't count the last double because we dont know if the pattern ended there. For this calculation, we count both: the number of islands, as well as sum their geometric average of the strainTimes between these transitions. This is done so we can nerf any repetitive patterns so as to not overbuff them, which is the purpose of the code below.

```cs
            double rhythmComplexitySum = 0.0;

            for (int i = 0; i < islandSizes.Length; i++)
            {
                if (islandSizes[i] != 0)
                    rhythmComplexitySum += islandTimes[i] / Math.Pow(islandSizes[i], .5); // sum the total amount of rhythm variance, penalizing for repeated island sizes.
            }
```

We also reward for any special transitions. These are transitions where you go from 1-3->1/4, or from 1/2-> 3/4. These are typically very difficult, so I decided to add a special bonus for these.

Now onto the main code.

```
            double strainValue = .25;

            double avgDeltaTime = (osuCurrent.StrainTime + Math.Max(50, Previous[0].DeltaTime)) / 2;

            double rhythmComplexity = calculateRhythmDifficulty(); // equals 1 with no rhythm difficulty, otherwise scales with a sqrt

            if (strainTimeBuffRange / avgDeltaTime > 1) // scale tap value for high BPM (above 200).
                strainValue += Math.Pow(strainTimeBuffRange / avgDeltaTime, 2);
            else
                strainValue += Math.Pow(strainTimeBuffRange / avgDeltaTime, 1);

            singleStrain *= computeDecay(baseDecay, osuCurrent.StrainTime);
            singleStrain += (.5 + osuCurrent.SnapProbability) * strainValue * singleMultiplier;

            currentStrain *= computeDecay(baseDecay, osuCurrent.StrainTime);
            currentStrain += strainValue * strainMultiplier;

            return Math.Max(Math.Min((1 / (1 - baseDecay)) * strainValue * strainMultiplier, // prevent over buffing strain past death stream level
                                      currentStrain * rhythmComplexity),
                            singleStrain); // we use a seperate strain for singles to not complete nuke boring 1-2 maps
 ```

The code is quite simple here, we take the average of the last two notes as our avgDeltaTime, and then scale if the strainTime is less than 75 Ms (or greater than 200BPM). Beyond that, we *currently* have two strain curves, one that is sensitive to single taps and one that is sensitive to rhythm. To be honest, I don't like this implementation. The idea here is to not overnerf maps with simple rhythms, but honestly... I can't see a reason to reward a map for having no challenging rhythm, but the feedback desired that there be something for these maps. We take the max of these two strain curves, and thats that.

### Accuracy 

With tr3's help, we updated the accuracy algorithm to be written with real variance in mind. One of the problems with the current curve is that playing on OD11 is unfairly brutal towards low accuracy. Tr3's method here calculates the real estimated variance needed to produce the accuracy of the play. This prevents cases where an OD9 HDDT SS is worth more accPP than a OD11 HDHRDT 95%, when the 95% actually requires greater accuracy. This does have drawbacks though, as it makes it more motivating to put a higher OD on a map so the total PP amount possible is greater, but I think this is already incentivized anyway

### AR 11 Changes (again)

I slightly changed the AR11 buffs to better balance with this new acc curve. I removed the length scaling on tap as well as it didn't make sense to me how AR11's difficulty correlated with tap. I also added a small flat buff to AR11 at 0 objects on aim, just to make short maps worth a little bit more.

### Speed Acc Curve Changes

I removed the very lenient acc curve at high BPMs since accuracy itself gives more pp now for the higher OD ranges. 

## Conclusion

I probably went into both too much and not enough detail at the same time in this PR. I'm taking a break from balancing for a couple weeks / months after this PR, so I wanted to try to explain as much as I could. I believe the values from this PR to be "good" and certainly more interesting than live, and wouldn't mind to see it deployed as is. If there are members of the community who are not pleased with these changes, or simply wish to improve values themselves, they are free to do so and create PRs for it. Balance is one of those things where the second you think you found it you look somewhere else and lose it. The current PP system isn't balanced. This one isn't either. It doesn't have to be.

One of the major restrictions I tried to keep myself held to with this build was to not make any attribute changes.I felt that it was best to leave those changes for a future patch and to do as much as I could without any disruptions to existing DB infrastructure.It is possible with several dozen new attributes that we could use full version of joz's SR calc to properly calculate non full combo plays.  This was probably a completely unneeded restriction, but it was fun.

If you would like to see results from this current build, you can find them at https://xexxar.stanr.info/.

If you want to subscribe to me on patreon, don't. I do this for 'fun'

Thanks for reading and taking the time to consider this implementation,

He Who Must Decrease (Xexxar)

### CODE REVIEW

**Comment by peppy** [comment]
On an initial check, your commit messages seem a bit all over the place. If possible, could you edit them (by rewriting your branch history) and force push once before we begin review? They don't need to be insanely detailed, but at least a short sentence explaining what the code change is, which some of them fail to capture.

**Comment by Xexxar** [comment]
> On an initial check, your commit messages seem a bit all over the place. If possible, could you edit them (by rewriting your branch history) and force push once before we begin review? They don't need to be insanely detailed, but at least a short sentence explaining what the code change is, which some of them fail to capture.
> 
> Note that you shouldn't be rebasing/combining/redating commits. Just editing the messages.

The more meme the commit, the more frustrated I was. I can go through and clean them up.

**Comment by hidraulico** [comment]
From a players perspective (and based from the values seen in xexxar.stanr.info), this needs quite a lot of work still. 

Stuff like [Remote Control](https://osu.ppy.sh/beatmapsets/351630#osu/774965)+DT which are 248bpm medium sized jumps with some bursts being worth tons more than [M.R.](https://osu.ppy.sh/beatmapsets/580867#osu/1229732)+DT which are 300bpm CS5 cutstreams doesn't feel right.

**Comment by peppy** [comment]
@hidraulico I'd highly recommend you read the full opening post. This is a code contribution aimed to open up the possibilities going forward. It does not need to be perfect and will continue to be adjusted as this is a live system we are working with.

This thread is a forum for discussion of the code quality and direction of the modularity. If we need subjective player opinions we will reach out elsewhere.

**Comment by hidraulico** [comment]
I am aware that this is very technical and in a very premature state. But I just wanna ask you peppy, to please reach out to more competent players (when the time is proper). This has been a clear issue in the past few reworks and no one has done anything about it.

File: osu.Game.Rulesets.Osu/Difficulty/Skills/OsuSkill.cs [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+using osu.Game.Rulesets.Osu.Difficulty.Preprocessing;
+using osu.Game.Rulesets.Osu.Objects;
+
+namespace osu.Game.Rulesets.Osu.Difficulty.Skills
+{
+    public abstract class OsuSkill : Skill
+    {
+        private readonly List<double> strains = new List<double>();
+        private readonly List<double> times = new List<double>();
+        private double target_fc_precision = 0.01;
```
Comment (blinchk) [inline osu.Game.Rulesets.Osu/Difficulty/Skills/OsuSkill.cs:19]: What is a reason to use snake_case, when you are using camelCase in another place for not const vars?

File: osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs [hunk 2]
```diff
@@ -149,15 +142,13 @@ private double computeSpeedValue()
...
-            speedValue *= 1.0 + Math.Min(approachRateFactor, approachRateFactor * (totalHits / 1000.0));
-
-            if (mods.Any(m => m is OsuModHidden))
-                speedValue *= 1.0 + 0.04 * (12.0 - Attributes.ApproachRate);
+            // scale speed sensitive to AR, without respect to object count.
+            speedValue *= 1.0 + approachRateFactor;
 
             // Scale the speed value with accuracy and OD
-            speedValue *= (0.95 + Math.Pow(Attributes.OverallDifficulty, 2) / 750) * Math.Pow(accuracy, (14.5 - Math.Max(Attributes.OverallDifficulty, 8)) / 2);
+            speedValue *= (0.575 + Math.Pow(Attributes.OverallDifficulty, 2) / 250) * Math.Pow(accuracy, 2.75);//(14.5 - Math.Max(Attributes.OverallDifficulty, 8)) / 2);
```
Comment (blinchk) [inline osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:151]: Looks like this variables can be too in consts to avoid magic numbers.

**Review by blinchk — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs [hunk 3]
```diff
@@ -166,25 +157,30 @@ private double computeSpeedValue()
...
+            double pm = (double)countMiss / amountHitObjectsWithAccuracy;
+            double p300 = 1.0 - pm - p100 - p50;
 
-            // It is possible to reach a negative accuracy with this formula. Cap it at zero - zero points
-            if (betterAccuracyPercentage < 0)
-                betterAccuracyPercentage = 0;
+            double m300 = 79.5 - 6.0 * Attributes.OverallDifficulty;
+            double m100 = 139.5 - 8.0 * Attributes.OverallDifficulty;
+            double m50 = 199.5 - 10.0 * Attributes.OverallDifficulty;
+            double acc = p300 + 1.0 / 3.0 * p100 + 1.0 / 6.0 * p50;
```
Comment (kpei) [inline osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:175]: var never used

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Tap.cs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+                        {
+                            islandTimes[islandSize] = islandTimes[islandSize] + 100.0 / Math.Sqrt(prevDelta * currDelta) * ((double)i / HistoryLength);
+                            islandSizes[islandSize] = islandSizes[islandSize] + 1;
+                        }
+
+                        islandSize = 0; // reset and count again, we sped up (usually this could only be if we did a 1/2 -> 1/3 -> 1/4) (or 1/1 -> 1/2 -> 1/4)
+                    }
+                    else // we're not the same or speeding up, must be slowing down.
+                    {
+                        if (islandSize > 6)
```
Comment (kpei) [inline osu.Game.Rulesets.Osu/Difficulty/Skills/Tap.cs:95]: duplicate of the above logic

File: osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs (hunk 3 above)
Comment (peppy) [reply osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:175]: Just as a heads-up, these kinds of issues are already pointed out in our style rules (and still need to be applied to this PR).

https://ci.appveyor.com/project/peppy/osu/builds/39360000

**Comment by bdach** [comment]
I can probably start reading through this at my own pace at some point within the next 2 weeks, but before I do, is this in a ready-to-review state, or are there further changes/adjustments coming?

**Comment by peppy** [comment]
@Xexxar when closing PRs please leave a message with reasoning for doing so to leave a paper trail.

**Comment by Xexxar** [comment]
> @Xexxar when closing PRs please leave a message with reasoning for doing so to leave a paper trail.

deleted due to lack of interest in supporting development in this area anymore.

**Pull request closed without merging**